node/hyperbahn/handler.js

```diff
@@ -0,0 +1,229 @@
+// THE SOFTWARE.
+
+'use strict';
+
+/* jshint maxparams:5 */
+
+var assert = require('assert');
+var util = require('util');
+
+var TChannelJSON = require('../as/json');
+var EgressNodes = require('./egress-nodes');
+var TChannelEndpointHandler = require('../endpoint-handler');
+
+module.exports = HyperbahnHandler;
+
+function HyperbahnHandler(options) {
+    if (!(this instanceof HyperbahnHandler)) {
+        return new HyperbahnHandler(options);
+    }
+
+    var self = this;
+    TChannelEndpointHandler.call(self, 'hyperbahn');
+
+    assert(options && options.channel, 'channel required');
+    assert(options && options.ringpop, 'ringpop required');
+    assert(options && options.egressNodes, 'egressNodes required');
+
+    // TODO support blackList
+
+    self.channel = options.channel;
+    self.type = 'hyperbahn.advertisement-handler';
+    self.ringpop = options.ringpop;
+    self.egressNodes = options.egressNodes;
+
+    self.tchannelJSON = TChannelJSON({
+        logger: self.channel.logger
+    });
+
+    // TODO replace JSON with real bufrw handlers for this
+    self.tchannelJSON.register(self, 'ad', self,
+        self.handleAdvertise);
+    self.tchannelJSON.register(self, 'relay-ad', self,
+        self.handleRelayAdvertise);
+}
+util.inherits(HyperbahnHandler, TChannelEndpointHandler);
+
+/*  req: {
+        services: Array<{
+            serviceName: String,
+            cost: Number
+        }>
+    }
+
+    res: {
+        connectionCount: Number
+    }
+*/
+HyperbahnHandler.prototype.handleAdvertise =
+function handleAdvertise(self, req, arg2, arg3, cb) {
+    /*eslint max-statements: [2, 25], max-params: [2, 5]*/
+    var services = arg3.services;
+
+    var servicesByExitNode = {};
+
+    for (var i = 0; i < services.length; i++) {
+        var serviceObj = services[i];
+        serviceObj.hostPort = req.remoteAddr;
+
+        var serviceName = serviceObj.serviceName;
+        if (serviceName === '') {
+            continue;
+        }
+
+        var exitNodes = self.egressNodes.exitsFor(serviceName);
+        var exitHosts = Object.keys(exitNodes);
+
+        for (var j = 0; j < exitHosts.length; j++) {
+            var exitNode = exitHosts[j];
+
+            var relayReq = servicesByExitNode[exitNode];
+            if (!relayReq) {
+                relayReq = servicesByExitNode[exitNode] = [];
+            }
+
+            relayReq.push(serviceObj);
+        }
+    }
+
+    var exitNodeKeys = Object.keys(servicesByExitNode);
+    var counter = 1 + exitNodeKeys.length;
+    for (var k = 0; k < exitNodeKeys.length; k++) {
+
+        var hostPort = exitNodeKeys[k];
+        var exitNodeServices = servicesByExitNode[hostPort];
+
+        self.sendRelayAdvertise(
+            hostPort, exitNodeServices, onFinish
+        );
+    }
+
+    onFinish();
+
+    // TODO remove blocking on fanout finish. Requires fixing
+    // autobahn tests upstream
+    function onFinish() {
+        if (--counter === 0) {
```

Comment:
I would stand by ===. Callback should not be called more than once. We should also verify that "send" guarantees at-most-once callback by discarding tardy extra results, otherwise, onFinish may count down earlier than intended.